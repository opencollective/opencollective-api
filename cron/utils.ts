import config from 'config';

/**
 * Heroku scheduler only has daily or hourly cron jobs, we only want to run
 * this script once per week on Monday (1). If the day is not Monday on production
 * we won't execute the script
 */
export function onlyExecuteInProdOnMondays() {
  const today = new Date();
  if (config.env === 'production' && today.getDay() !== 1) {
    console.log('OC_ENV is production and day is not Monday, script aborted!');
    process.exit(0);
  }
}
