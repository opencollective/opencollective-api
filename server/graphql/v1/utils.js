import debugLib from 'debug';
import { graphql } from 'graphql';

import { loaders } from '../loaders';

import schema from './schema';

const debug = debugLib('graphql');

export const makeRequest = (remoteUser, query) => {
  return {
    remoteUser,
    body: { query },
    loaders: loaders({ remoteUser }),
  };
};

export const graphqlQuery = async (query, variables, remoteUser) => {
  const prepare = () => {
    if (remoteUser) {
      remoteUser.rolesByCollectiveId = null; // force refetching the roles
      return remoteUser.populateRoles();
    } else {
      return Promise.resolve();
    }
  };

  if (process.env.DEBUG && process.env.DEBUG.match(/graphql/)) {
    debug('query', query);
    debug('variables', variables);
    debug('context', remoteUser);
  }

  return prepare().then(() =>
    graphql({
      schema,
      source: query,
      rootValue: null,
      contextValue: makeRequest(remoteUser, query),
      variableValues: variables,
    }),
  );
};
