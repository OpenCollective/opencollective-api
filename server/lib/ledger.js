/** @module lib/ledger */

import models from '../models';

export async function rows(TransactionGroup) {
  return models.Transaction.findAll({
    where: { TransactionGroup }
  });
}

/** Aggregate amount fields of array of transactions
 *
 * @param {number} id is the id of the ledger's Collective that will
 *  have the balance formatted. This works for any type of collective:
 *  Users, Collectives, & Hosts.
 * @param {models.Transaction[]} rows is the array of transaction rows
 *  that the balance will be calculated from.
 * @return {Object} with currencies as keys and amounts as values.
 * @example
 * > const rows = await libledger.rows('7300e56b-09c4-4e8a-8c11-cd17cf7fdd3b');
 * > balance(userCollective.id, rows)
 * { USD: 50, MXN: 922 }
 */
export function balance(id, rows) {
  const { currency, fromCurrency } = rows[0].dataValues;
  const initial = { [currency]: 0, [fromCurrency]: 0 };
  return rows
    .filter(x => x.CollectiveId === id)
    .reduce((a, b) => {
      a[currency] += b.amount;
      // Prevent from accounting for the same value twice
      if (currency !== fromCurrency) {
        a[fromCurrency] += b.fromAmount;
      }
      return a;
    }, initial);
}

/**
 * Format the balance of the "left side" of a collective ledger
 *
 * This function will aggregate the value of the field `amount` in the
 * `Transactions` table. Which means that it will use whatever
 * currency was accepted by the host when the transaction was created.
 *
 * @param {number} id is the id of the ledger's Collective that will
 *  have the balance formatted. This works for any type of collective:
 *  Users, Collectives, & Hosts.
 * @param {models.Transaction[]} rows is the array of transaction rows
 *  that the balance will be calculated from.
 * @return {string} with value and currency concatenated.
 */
export function formattedBalance(id, rows) {
  const { currency } = rows[0].dataValues;
  const sum = balance(id, rows);
  return `${sum[currency]} ${currency}`;
}

/**
 * Format the balance of the "left side" of a User's collective ledger
 *
 * This function will aggregate the value of the field `fromAmount` in
 * the `Transactions` table. Which means that *it will use whatever
 * currency the User used when the transaction was created*.
 *
 * @param {number} id is the id of the ledger's Collective that will
 *  have the balance formatted. This works for any type of collective:
 *  Users, Collectives, & Hosts.
 * @param {models.Transaction[]} rows is the array of transaction rows
 *  that the balance will be calculated from.
 * @return {string} with value and currency concatenated.
 */
export function formattedBalanceFromCurrency(id, rows) {
  const { fromCurrency } = rows[0].dataValues;
  const sum = balance(id, rows);
  return `${sum[fromCurrency]} ${fromCurrency}`;
}
