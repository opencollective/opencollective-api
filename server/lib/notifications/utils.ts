import sanitizeHtml from 'sanitize-html';

import { sanitizerOptions as updateSanitizerOptions } from '../../models/Update';

const constructPreviewImageURL = (service: string, id: string) => {
  if (service === 'youtube' && id.match('[a-zA-Z0-9_-]{11}')) {
    return `https://img.youtube.com/vi/${id}/0.jpg`;
  } else if (service === 'anchorFm') {
    return `https://opencollective.com/static/images/anchor-fm-logo.png`;
  } else {
    return null;
  }
};

const parseServiceLink = (videoLink: string) => {
  const regexps = {
    youtube: new RegExp(
      '(?:https?://)?(?:www\\.)?youtu(?:\\.be/|be(-nocookie)?\\.com/\\S*(?:watch|embed)(?:(?:(?=/[^&\\s?]+(?!\\S))/)|(?:\\S*v=|v/)))([^&\\s?]+)',
      'i',
    ),
    anchorFm: /^(http|https)?:\/\/(www\.)?anchor\.fm\/([^/]+)(\/embed)?(\/episodes\/)?([^/]+)?\/?$/,
  };
  for (const service in regexps) {
    videoLink = videoLink.replace('/?showinfo=0', '');
    const matches = regexps[service].exec(videoLink);
    if (matches) {
      if (service === 'anchorFm') {
        const podcastName = matches[3];
        const episodeId = matches[6];
        const podcastUrl = `${podcastName}/embed`;
        return { service, id: episodeId ? `${podcastUrl}/episodes/${episodeId}` : podcastUrl };
      } else {
        return { service, id: matches[matches.length - 1] };
      }
    }
  }
  return {};
};

export const replaceVideosByImagePreviews = (html: string) => {
  const sanitizerOptions = {
    ...updateSanitizerOptions,
    transformTags: {
      ...updateSanitizerOptions.transformTags,
      iframe: (tagName, attribs) => {
        if (!attribs.src) {
          return '';
        }
        const { service, id } = parseServiceLink(attribs.src);
        const imgSrc = constructPreviewImageURL(service, id);
        if (imgSrc) {
          return {
            tagName: 'img',
            attribs: {
              src: imgSrc,
              alt: `${service} content`,
            },
          };
        } else {
          return '';
        }
      },
    },
  };
  return sanitizeHtml(html, sanitizerOptions);
};
