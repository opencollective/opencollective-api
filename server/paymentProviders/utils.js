import slugify from 'limax';

import activities from '../constants/activities';
import { CollectiveType } from '../constants/collectives';
import ExpenseStatus from '../constants/expense-status';
import ExpenseType from '../constants/expense-type';
import { PAYMENT_METHOD_SERVICE, PAYMENT_METHOD_TYPE } from '../constants/paymentMethods';
import { TransactionKind } from '../constants/transaction-kind';
import { TransactionTypes } from '../constants/transactions';
import { getFxRate } from '../lib/currency';
import { crypto } from '../lib/encryption';
import logger from '../lib/logger';
import { toNegative } from '../lib/math';
import { createRefundTransaction } from '../lib/payments';
import models, { Op } from '../models';

export const getVirtualCardForTransaction = async cardId => {
  return models.VirtualCard.findOne({
    where: {
      id: cardId,
    },
    include: [
      { association: 'collective', required: true },
      { association: 'host', required: true },
      { association: 'user' },
    ],
  });
};

export const notifyCollectiveMissingReceipt = async (expense, virtualCard) => {
  expense.collective = expense.collective || (await expense.getCollective());
  expense.fromCollective = expense.fromCollective || (await expense.getFromCollective());
  virtualCard = virtualCard || expense.virtualCard || (await expense.getVirtualCard());

  if (expense.collective.settings?.ignoreExpenseMissingReceiptAlerts === true) {
    return;
  }

  expense.createActivity(
    activities.COLLECTIVE_EXPENSE_MISSING_RECEIPT,
    { id: virtualCard.UserId },
    { ...expense.data, user: virtualCard.user },
  );
};

export const persistTransaction = async (virtualCard, transaction) => {
  // Make sure amount is an absolute value
  const amount = Math.abs(transaction.amount);

  if (amount === 0) {
    return;
  }

  const transactionId = transaction.id;
  const expenseData = { transactionId };
  const UserId = virtualCard.UserId;
  const host = virtualCard.host;
  const collective = virtualCard.collective;
  const vendor = await getOrCreateVendor(transaction.vendorProviderId, transaction.vendorName);
  const currency = transaction.currency || 'USD';
  const hostCurrencyFxRate = await getFxRate(currency, host.currency);
  const description = `Virtual Card charge: ${vendor.name}`;
  const paymentMethod = await host.findOrCreatePaymentMethod(
    PAYMENT_METHOD_SERVICE.STRIPE,
    PAYMENT_METHOD_TYPE.VIRTUAL_CARD,
  );

  // Case when expense is already created after the stripe authorization request event
  // Double check if transaction is a refund because sometimes the refund event is sent before the charge event
  if (transaction.fromAuthorizationId && !transaction.isRefund) {
    const processingExpense = await models.Expense.findOne({
      where: {
        status: ExpenseStatus.PROCESSING,
        VirtualCardId: virtualCard.id,
        data: { authorizationId: transaction.fromAuthorizationId },
      },
    });

    if (processingExpense) {
      // Make sure we update the Expense and ExpenseItem amounts.
      // Sometimes there's a difference between the authorized amount and the charged amount.
      await models.ExpenseItem.update({ amount }, { where: { ExpenseId: processingExpense.id } });
      await processingExpense.update({
        amount,
        paymentMethodId: paymentMethod.id,
        data: { ...expenseData, missingDetails: true, ...transaction.data },
      });
      // Mark Expense as Paid, create activity and don't send notifications
      await processingExpense.markAsPaid({ skipActivity: true });

      await models.Transaction.createDoubleEntry({
        CollectiveId: collective.id,
        FromCollectiveId: vendor.id,
        HostCollectiveId: host.id,
        description,
        type: 'DEBIT',
        currency,
        ExpenseId: processingExpense.id,
        PaymentMethodId: paymentMethod.id,
        amount: toNegative(amount),
        netAmountInCollectiveCurrency: toNegative(amount),
        hostCurrency: host.currency,
        amountInHostCurrency: Math.round(toNegative(amount) * hostCurrencyFxRate),
        paymentProcessorFeeInHostCurrency: 0,
        hostFeeInHostCurrency: 0,
        platformFeeInHostCurrency: 0,
        hostCurrencyFxRate,
        kind: TransactionKind.EXPENSE,
        data: transaction.data,
        clearedAt: transaction.clearedAt,
      });

      const expenseAttachment = await models.ExpenseAttachedFile.findOne({
        where: { ExpenseId: processingExpense.id },
      });

      const expenseItemWithAttachment = await models.ExpenseItem.findOne({
        where: { ExpenseId: processingExpense.id, url: { [Op.ne]: null } },
      });

      if (!expenseAttachment && !expenseItemWithAttachment) {
        await notifyCollectiveMissingReceipt(processingExpense, virtualCard);
      }

      return processingExpense;
    }
  }

  const existingExpense = await models.Expense.findOne({
    where: {
      status: ExpenseStatus.PAID,
      VirtualCardId: virtualCard.id,
      // TODO : only let transactionId in a few months (today : 11/2021) or make a migration to update data on existing expenses and transactions
      data: { [Op.or]: [{ transactionId }, { id: transactionId }, { token: transactionId }] },
    },
  });

  if (existingExpense) {
    logger.warn(`Virtual Card charge already reconciled, ignoring it: ${transactionId}`);
    return;
  }

  if (transaction.isRefund) {
    const existingTransaction = await models.Transaction.findOne({
      where: {
        CollectiveId: collective.id,
        // TODO : only let refundTransactionId in a few months (today : 11/2021) or make a migration to update data on existing expenses and transactions
        data: {
          [Op.or]: [{ refundTransactionId: transactionId }, { id: transactionId }, { token: transactionId }],
        },
      },
    });

    if (existingTransaction) {
      logger.warn(`Virtual Card refund already reconciled, ignoring it: ${transactionId}`);
      return;
    }

    const expense = await models.Expense.findOne({
      where: {
        CollectiveId: collective.id,
        FromCollectiveId: vendor.id,
        VirtualCardId: virtualCard.id,
        type: ExpenseType.CHARGE,
        status: ExpenseStatus.PAID,
        amount,
      },
      order: [['createdAt', 'DESC']],
    });

    if (expense) {
      const [originalCreditTransaction] = await expense.getTransactions({
        where: { type: TransactionTypes.CREDIT, kind: TransactionKind.EXPENSE },
      });
      if (originalCreditTransaction?.amount === amount) {
        if (!originalCreditTransaction.RefundTransactionId) {
          await createRefundTransaction(originalCreditTransaction, 0, { refundTransactionId: transactionId });
        }
        return;
      }
    }

    await models.Transaction.createDoubleEntry({
      CollectiveId: collective.id,
      FromCollectiveId: vendor.id,
      HostCollectiveId: host.id,
      description: `Virtual Card refund: ${vendor.name}`,
      type: 'CREDIT',
      currency,
      amount: amount,
      netAmountInCollectiveCurrency: amount,
      hostCurrency: host.currency,
      amountInHostCurrency: Math.round(amount * hostCurrencyFxRate),
      paymentProcessorFeeInHostCurrency: 0,
      hostFeeInHostCurrency: 0,
      platformFeeInHostCurrency: 0,
      hostCurrencyFxRate,
      isRefund: true,
      kind: TransactionKind.EXPENSE,
      ExpenseId: expense?.id,
      PaymentMethodId: paymentMethod.id,
      data: { refundTransactionId: transactionId },
    });

    return;
  }

  let expense;

  try {
    expense = await models.Expense.create({
      UserId,
      CollectiveId: collective.id,
      FromCollectiveId: vendor.id,
      HostCollectiveId: host.id,
      currency,
      amount,
      description,
      VirtualCardId: virtualCard.id,
      PaymentMethodId: paymentMethod.id,
      lastEditedById: UserId,
      status: ExpenseStatus.PAID,
      type: ExpenseType.CHARGE,
      incurredAt: transaction.incurredAt,
      data: { ...expenseData, missingDetails: true },
    });

    await models.ExpenseItem.create({
      ExpenseId: expense.id,
      incurredAt: transaction.incurredAt,
      CreatedByUserId: UserId,
      amount,
      currency,
    });

    await models.Transaction.createDoubleEntry({
      CollectiveId: collective.id,
      FromCollectiveId: vendor.id,
      HostCollectiveId: host.id,
      description,
      type: 'DEBIT',
      currency,
      ExpenseId: expense.id,
      PaymentMethodId: paymentMethod.id,
      amount: toNegative(amount),
      netAmountInCollectiveCurrency: toNegative(amount),
      hostCurrency: host.currency,
      amountInHostCurrency: Math.round(toNegative(amount) * hostCurrencyFxRate),
      paymentProcessorFeeInHostCurrency: 0,
      hostFeeInHostCurrency: 0,
      platformFeeInHostCurrency: 0,
      hostCurrencyFxRate,
      kind: TransactionKind.EXPENSE,
    });

    await notifyCollectiveMissingReceipt(expense, virtualCard);

    return expense;
  } catch (e) {
    if (expense) {
      await models.Transaction.destroy({ where: { ExpenseId: expense.id } });
      await models.ExpenseItem.destroy({ where: { ExpenseId: expense.id } });
      await expense.destroy();
    }
    throw e;
  }
};

export const getOrCreateVendor = async (vendorProviderId, vendorName) => {
  const slug = vendorProviderId.toString().toLowerCase();
  const hash = crypto.hash(`${slug}${vendorName}`);
  const uniqueSlug = slugify(`${vendorName}-${hash.slice(0, 6)}`);

  const [vendor] = await models.Collective.findOrCreate({
    where: { [Op.or]: [{ slug }, { slug: uniqueSlug }] },
    defaults: { name: vendorName, type: CollectiveType.VENDOR, slug: uniqueSlug },
  });

  if (vendor.name !== vendorName) {
    logger.warn(`Virtual Card: vendor name mismatch for ${vendorProviderId}: '${vendorName}' / '${vendor.name}'`);
  }
  // Update existing vendor to use uniqueSlug.
  if (vendor.slug === slug) {
    await vendor.update({ slug: uniqueSlug });
  }

  return vendor;
};
