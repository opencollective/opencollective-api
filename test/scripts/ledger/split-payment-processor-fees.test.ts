import { main as splitPaymentProcessorFees } from '../../../scripts/ledger/split-payment-processor-fees';
import { refundTransaction } from '../../../server/lib/payments';
import { fakeCollective, fakeHost, fakeTransaction, fakeUser } from '../../test-helpers/fake-data';
import { resetTestDB, seedDefaultVendors, snapshotLedger } from '../../utils';

const SNAPSHOT_COLUMNS = [
  'kind',
  'type',
  'amount',
  'paymentProcessorFeeInHostCurrency',
  'netAmountInCollectiveCurrency',
  'CollectiveId',
  'FromCollectiveId',
  'HostCollectiveId',
  'isRefund',
];

describe('scripts/ledger/split-payment-processor-fees', () => {
  beforeEach('reset test database', async () => {
    await resetTestDB();
    await seedDefaultVendors();
  });

  it('1. migrate a regular contribution with payment processor fees', async () => {
    const user = await fakeUser(null, { name: 'User' });
    const host = await fakeHost({ name: 'Host' });
    const collective = await fakeCollective({ name: 'Collective', HostCollectiveId: host.id });
    await fakeTransaction(
      {
        amount: 500,
        kind: 'CONTRIBUTION',
        paymentProcessorFeeInHostCurrency: -25,
        FromCollectiveId: user.CollectiveId,
        CollectiveId: collective.id,
        HostCollectiveId: host.id,
      },
      { createDoubleEntry: true },
    );
    await snapshotLedger(SNAPSHOT_COLUMNS);

    await splitPaymentProcessorFees('migrate');

    await snapshotLedger(SNAPSHOT_COLUMNS);
  });

  it('2. migrate a regular refunded contribution with payment processor fees', async () => {
    const user = await fakeUser(null, { name: 'User' });
    const host = await fakeHost({ name: 'Host' });
    const collective = await fakeCollective({ name: 'Collective', HostCollectiveId: host.id });
    const transaction = await fakeTransaction(
      {
        amount: 500,
        kind: 'CONTRIBUTION',
        paymentProcessorFeeInHostCurrency: -25,
        FromCollectiveId: user.CollectiveId,
        CollectiveId: collective.id,
        HostCollectiveId: host.id,
      },
      { createDoubleEntry: true },
    );

    // Refund the transaction
    await refundTransaction(transaction, user);
    await snapshotLedger(SNAPSHOT_COLUMNS);

    await splitPaymentProcessorFees('migrate');

    await snapshotLedger(SNAPSHOT_COLUMNS);
  });

  it('3. migrate an expense with payment processor fees', async () => {
    const user = await fakeUser(null, { name: 'User' });
    const host = await fakeHost({ name: 'Host' });
    const collective = await fakeCollective({ name: 'Collective', HostCollectiveId: host.id });

    await fakeTransaction(
      {
        amount: -500,
        kind: 'EXPENSE',
        paymentProcessorFeeInHostCurrency: -25,
        netAmountInCollectiveCurrency: -525,
        currency: 'USD',
        hostCurrency: 'USD',
        FromCollectiveId: user.CollectiveId,
        CollectiveId: collective.id,
        HostCollectiveId: host.id,
      },
      { createDoubleEntry: true },
    );

    await snapshotLedger(SNAPSHOT_COLUMNS);

    await splitPaymentProcessorFees('migrate');

    await snapshotLedger(SNAPSHOT_COLUMNS);
  });

  it('4. migrate an expense with payment processor fees AND taxes', async () => {
    const user = await fakeUser(null, { name: 'User' });
    const host = await fakeHost({ name: 'Host' });
    const collective = await fakeCollective({ name: 'Collective', HostCollectiveId: host.id });
    await fakeTransaction(
      {
        amount: -500,
        kind: 'EXPENSE',
        paymentProcessorFeeInHostCurrency: -25,
        netAmountInCollectiveCurrency: -675,
        currency: 'USD',
        hostCurrency: 'USD',
        FromCollectiveId: user.CollectiveId,
        CollectiveId: collective.id,
        taxAmount: -150,
        HostCollectiveId: host.id,
      },
      { createDoubleEntry: true },
    );
    await snapshotLedger([...SNAPSHOT_COLUMNS, 'taxAmount']);
    await splitPaymentProcessorFees('migrate');
    await snapshotLedger([...SNAPSHOT_COLUMNS, 'taxAmount']);
  });

  // select * from "Transactions" where kind = 'EXPENSE' AND data ->> 'feesPayer' = 'PAYEE' LIMIT 100
  it('5. migrate an expense with processor fees on the payee if feesPayer=PAYEE', async () => {
    const user = await fakeUser(null, { name: 'User' });
    const host = await fakeHost({ name: 'Host' });
    const collective = await fakeCollective({ name: 'Collective', HostCollectiveId: host.id });
    await fakeTransaction(
      {
        amount: -475,
        kind: 'EXPENSE',
        paymentProcessorFeeInHostCurrency: -25,
        netAmountInCollectiveCurrency: -500,
        currency: 'USD',
        hostCurrency: 'USD',
        FromCollectiveId: user.CollectiveId,
        CollectiveId: collective.id,
        HostCollectiveId: host.id,
        data: { feesPayer: 'PAYEE' },
      },
      { createDoubleEntry: true },
    );
    await snapshotLedger([...SNAPSHOT_COLUMNS, 'taxAmount']);
    await splitPaymentProcessorFees('migrate');
    await snapshotLedger([...SNAPSHOT_COLUMNS, 'taxAmount']);
  });

  // TODO
  //
  // it('handles multi-currency contributions with refunds', () => {})
  // it('handles multi-currency expenses with refunds', () => {})
});
