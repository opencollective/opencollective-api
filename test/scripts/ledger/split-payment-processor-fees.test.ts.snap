// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scripts/ledger/split-payment-processor-fees 1. migrate a regular contribution with payment processor fees 1`] = `
"
| kind         | type   | amount | paymentFee | netAmountInCollectiveCurrency | To         | From       | Host | isRefund |
| ------------ | ------ | ------ | ---------- | ----------------------------- | ---------- | ---------- | ---- | -------- |
| CONTRIBUTION | CREDIT | 500    | -25        | 500                           | Collective | User       | Host | false    |
| CONTRIBUTION | DEBIT  | -500   | -25        | -500                          | User       | Collective | NULL | false    |"
`;

exports[`scripts/ledger/split-payment-processor-fees 1. migrate a regular contribution with payment processor fees 2`] = `
"
| kind                  | type   | amount | paymentFee | netAmountInCollectiveCurrency | To                      | From                    | Host | isRefund |
| --------------------- | ------ | ------ | ---------- | ----------------------------- | ----------------------- | ----------------------- | ---- | -------- |
| PAYMENT_PROCESSOR_FEE | CREDIT | 25     | 0          | 25                            | Other Payment Processor | Collective              | NULL | false    |
| PAYMENT_PROCESSOR_FEE | DEBIT  | -25    | 0          | -25                           | Collective              | Other Payment Processor | Host | false    |
| CONTRIBUTION          | CREDIT | 500    | 0          | 500                           | Collective              | User                    | Host | false    |
| CONTRIBUTION          | DEBIT  | -500   | 0          | -500                          | User                    | Collective              | NULL | false    |"
`;

exports[`scripts/ledger/split-payment-processor-fees 2. migrate a regular refunded contribution with payment processor fees 1`] = `
"
| kind                    | type   | amount | paymentFee | netAmountInCollectiveCurrency | To         | From       | Host | isRefund |
| ----------------------- | ------ | ------ | ---------- | ----------------------------- | ---------- | ---------- | ---- | -------- |
| CONTRIBUTION            | CREDIT | 500    | 0          | 500                           | User       | Collective | NULL | true     |
| CONTRIBUTION            | DEBIT  | -500   | 0          | -500                          | Collective | User       | Host | true     |
| PAYMENT_PROCESSOR_COVER | CREDIT | 25     | 0          | 25                            | Collective | Host       | Host | true     |
| PAYMENT_PROCESSOR_COVER | DEBIT  | -25    | 0          | -25                           | Host       | Collective | Host | true     |
| CONTRIBUTION            | CREDIT | 500    | -25        | 500                           | Collective | User       | Host | false    |
| CONTRIBUTION            | DEBIT  | -500   | -25        | -500                          | User       | Collective | NULL | false    |"
`;

exports[`scripts/ledger/split-payment-processor-fees 2. migrate a regular refunded contribution with payment processor fees 2`] = `
"
| kind                    | type   | amount | paymentFee | netAmountInCollectiveCurrency | To                      | From                    | Host | isRefund |
| ----------------------- | ------ | ------ | ---------- | ----------------------------- | ----------------------- | ----------------------- | ---- | -------- |
| PAYMENT_PROCESSOR_FEE   | CREDIT | 25     | 0          | 25                            | Other Payment Processor | Collective              | NULL | false    |
| PAYMENT_PROCESSOR_FEE   | DEBIT  | -25    | 0          | -25                           | Collective              | Other Payment Processor | Host | false    |
| CONTRIBUTION            | CREDIT | 500    | 0          | 500                           | User                    | Collective              | NULL | true     |
| CONTRIBUTION            | DEBIT  | -500   | 0          | -500                          | Collective              | User                    | Host | true     |
| PAYMENT_PROCESSOR_COVER | CREDIT | 25     | 0          | 25                            | Collective              | Host                    | Host | true     |
| PAYMENT_PROCESSOR_COVER | DEBIT  | -25    | 0          | -25                           | Host                    | Collective              | Host | true     |
| CONTRIBUTION            | CREDIT | 500    | 0          | 500                           | Collective              | User                    | Host | false    |
| CONTRIBUTION            | DEBIT  | -500   | 0          | -500                          | User                    | Collective              | NULL | false    |"
`;

exports[`scripts/ledger/split-payment-processor-fees 3. migrate an expense with payment processor fees 1`] = `
"
| kind    | type   | amount | paymentFee | netAmountInCollectiveCurrency | To         | From       | Host | isRefund |
| ------- | ------ | ------ | ---------- | ----------------------------- | ---------- | ---------- | ---- | -------- |
| EXPENSE | CREDIT | 525    | -25        | 500                           | User       | Collective | NULL | false    |
| EXPENSE | DEBIT  | -500   | -25        | -525                          | Collective | User       | Host | false    |"
`;

exports[`scripts/ledger/split-payment-processor-fees 3. migrate an expense with payment processor fees 2`] = `
"
| kind                  | type   | amount | paymentFee | netAmountInCollectiveCurrency | To                      | From                    | Host | isRefund |
| --------------------- | ------ | ------ | ---------- | ----------------------------- | ----------------------- | ----------------------- | ---- | -------- |
| PAYMENT_PROCESSOR_FEE | CREDIT | 25     | 0          | 25                            | Other Payment Processor | Collective              | NULL | false    |
| PAYMENT_PROCESSOR_FEE | DEBIT  | -25    | 0          | -25                           | Collective              | Other Payment Processor | Host | false    |
| EXPENSE               | CREDIT | 500    | 0          | 500                           | User                    | Collective              | NULL | false    |
| EXPENSE               | DEBIT  | -500   | 0          | -500                          | Collective              | User                    | Host | false    |"
`;

exports[`scripts/ledger/split-payment-processor-fees 4. migrate an expense with payment processor fees AND taxes 1`] = `
"
| kind    | type   | amount | paymentFee | netAmountInCollectiveCurrency | To         | From       | Host | isRefund | tax  |
| ------- | ------ | ------ | ---------- | ----------------------------- | ---------- | ---------- | ---- | -------- | ---- |
| EXPENSE | CREDIT | 675    | -25        | 500                           | User       | Collective | NULL | false    | -150 |
| EXPENSE | DEBIT  | -500   | -25        | -675                          | Collective | User       | Host | false    | -150 |"
`;

exports[`scripts/ledger/split-payment-processor-fees 4. migrate an expense with payment processor fees AND taxes 2`] = `
"
| kind                  | type   | amount | paymentFee | netAmountInCollectiveCurrency | To                      | From                    | Host | isRefund | tax  |
| --------------------- | ------ | ------ | ---------- | ----------------------------- | ----------------------- | ----------------------- | ---- | -------- | ---- |
| PAYMENT_PROCESSOR_FEE | CREDIT | 25     | 0          | 25                            | Other Payment Processor | Collective              | NULL | false    |      |
| PAYMENT_PROCESSOR_FEE | DEBIT  | -25    | 0          | -25                           | Collective              | Other Payment Processor | Host | false    |      |
| EXPENSE               | CREDIT | 650    | 0          | 500                           | User                    | Collective              | NULL | false    | -150 |
| EXPENSE               | DEBIT  | -500   | 0          | -650                          | Collective              | User                    | Host | false    | -150 |"
`;

exports[`scripts/ledger/split-payment-processor-fees 5. migrate an expense with processor fees on the payee if feesPayer=PAYEE 1`] = `
"
| kind    | type   | amount | paymentFee | netAmountInCollectiveCurrency | To         | From       | Host | isRefund | tax |
| ------- | ------ | ------ | ---------- | ----------------------------- | ---------- | ---------- | ---- | -------- | --- |
| EXPENSE | CREDIT | 500    | -25        | 475                           | User       | Collective | NULL | false    |     |
| EXPENSE | DEBIT  | -475   | -25        | -500                          | Collective | User       | Host | false    |     |"
`;

exports[`scripts/ledger/split-payment-processor-fees 5. migrate an expense with processor fees on the payee if feesPayer=PAYEE 2`] = `
"
| kind                  | type   | amount | paymentFee | netAmountInCollectiveCurrency | To                      | From                    | Host | isRefund | tax |
| --------------------- | ------ | ------ | ---------- | ----------------------------- | ----------------------- | ----------------------- | ---- | -------- | --- |
| PAYMENT_PROCESSOR_FEE | CREDIT | 25     | 0          | 25                            | Other Payment Processor | Collective              | NULL | false    |     |
| PAYMENT_PROCESSOR_FEE | DEBIT  | -25    | 0          | -25                           | Collective              | Other Payment Processor | Host | false    |     |
| EXPENSE               | CREDIT | 475    | 0          | 475                           | User                    | Collective              | NULL | false    |     |
| EXPENSE               | DEBIT  | -475   | 0          | -475                          | Collective              | User                    | Host | false    |     |"
`;
